File: ruby/structure/heap.rb
class Heap
  attr_reader :arr

  def initialize(values = [])
    @arr = values

    (arr.size / 2 - 1).downto(0).each do |i|
      heapify!(i)
    end
  end

  def push(value)
    i = arr.size
    arr << value
    parent_i = (i - 1) / 2

    while i > 0 && arr[parent_i] < arr[i]
      swap(i, parent_i)
      i = parent_i
      parent_i = (i - 1) / 2
    end
  end

  def max
    arr.first
  end

  def pop
    value = arr[0]
    last  = arr.pop
    unless arr.empty?
      arr[0] = last
      heapify!(0)
    end
    value
  end

  def length
    arr.size
  end

  def merge(heap2)
    while heap2.max
      push(heap2.pop)
    end
  end

  private def heapify!(i)
    loop do
      left_child  = 2 * i + 1
      right_child = 2 * i + 2
      largest_child = i

      largest_child = left_child  if left_child  < arr.size && arr[left_child]  > arr[largest_child]
      largest_child = right_child if right_child < arr.size && arr[right_child] > arr[largest_child]

      break if largest_child == i

      swap(i, largest_child)
      i = largest_child
    end
  end

  private def swap(i, j)
    arr[i], arr[j] = arr[j], arr[i]
  end
end

# arr = 10.times.map { rand(0..9) }
# arr1 = arr.shuffle
# arr2 = arr.shuffle
# arr3 = arr.shuffle

# heap1 = Heap.new(arr1)
# heap2 = Heap.new(arr2)
# heap3 = Heap.new(arr3)
# heap4 = Heap.new; arr.each { |x| heap4.push(x) }
# heap5 = Heap.new; arr.reverse.each { |x| heap5.push(x) }

# heaps = [heap1, heap2, heap3, heap4, heap5]

# fail unless heaps.map(&:max).uniq == [arr.max]

# while heaps.first.length > 0
#   fail unless heaps.map(&:max).uniq.size == 1
#   fail unless heaps.map(&:max).uniq.size == 1
#   fail unless heaps.map(&:pop).uniq.size == 1
# end

File: javascript/structure/Heap.js
class Heap {
  constructor(values = [], compare = (a, b) => a - b) {
    this.data = values
    this.compare = compare

    for (let i = Math.floor(this.size / 2) - 1; i >= 0; --i) {
      this.#heapify(i)
    }
  }

  push(value) {
    this.data.push(value)

    let i = this.size - 1
    let parentI = this.#parentIndex(i)

    while (i > 0 && this.compare(this.data[parentI], this.data[i]) < 0) {
      this.#swap(i, parentI)
      i = parentI
      parentI = this.#parentIndex(i)
    }
  }

  get top() {
    return this.data[0]
  }

  pop() {
    const value = this.top
    const last  = this.data.pop()
    if (this.size > 0) {
      this.data[0] = last
      this.#heapify(0)
    }
    return value
  }

  get size() {
    return this.data.length
  }

  #parentIndex(i) {
    return Math.floor((i - 1) / 2)
  }

  #swap(i, j) {
    [this.data[i], this.data[j]] = [this.data[j], this.data[i]]
  }

  #heapify(i) {
    while (true) {
      const leftChild  = 2 * i + 1
      const rightChild = 2 * i + 2
      let largestChild = i

      if (leftChild < this.size && this.compare(this.data[largestChild], this.data[leftChild]) < 0) {
        largestChild = leftChild
      }
      if (rightChild < this.size && this.compare(this.data[largestChild], this.data[rightChild]) < 0) {
        largestChild = rightChild
      }

      if (largestChild === i) {
        break
      }

      this.#swap(i, largestChild)
      i = largestChild
    }
  }
}

// const arr1 = [8, 9, 5, 2, 1, 6, 7, 4, 10, 3]
// const arr2 = [1, 6, 8, 9, 3, 5, 2, 10, 4, 7]
// const arr3 = [6, 3, 8, 5, 9, 10, 4, 2, 7, 1]

// const heap1 = new Heap(arr1)
// const heap2 = new Heap(arr2)
// const heap3 = new Heap()
// arr3.forEach(x => heap3.push(x))

// const heaps = [heap1, heap2, heap3]
// console.log(heap1.data)
// console.log(heap3.data)

// heaps.forEach(heap => console.assert(heap.top === 10))

// while (heaps[0].size > 0) {
//   heaps.forEach(heap => console.assert(heap.top === heaps[0].size))
//   console.assert(new Set(heaps.map(heap => heap.pop())).size === 1)
// }
